File: Payload_Type/odin/mythic/agent_functions/bash_executor.py
from mythic_payloadtype_container.MythicCommandBase import *
import json
from mythic_payloadtype_container.MythicRPC import *


class BashArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = []

    async def parse_arguments(self):
        pass

class BashCommand(CommandBase):
    cmd = "bash_executor"
    needs_admin = False
    help_cmd = "bash {command}"
    description = "Execute a shell command using 'bash -c'"
    version = 1
    author = "@antman1p"
    attackmapping = ["T1059", "T1059.004"]
    argument_class = BashArguments
    attributes = CommandAttributes(
        suggested_command=True,
        supported_os=[SupportedOS.MacOS, SupportedOS.Linux]

    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        resp = await MythicRPC().execute("create_artifact", task_id=task.id,
            artifact="/bin/bash -c {}".format(task.args.command_line),
            artifact_type="Process Create",
        )
        resp = await MythicRPC().execute("create_artifact", task_id=task.id,
            artifact="{}".format(task.args.command_line),
            artifact_type="Process Create",
        )
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/freyja/freyja/agentfunctions/bash_executor.go
package agentfunctions

import (
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
)

var bash_executor = agentstructs.Command{
	Name:                "bash_executor",
	Author:              "@antman1p",
	Description:         "Execute a shell command using 'bash -c'",
	MitreAttackMappings: []string{"T1059"},
	CommandAttributes: agentstructs.CommandAttribute{
		CommandIsSuggested: true,
		CommandIsBuiltin:   false,
		SupportedOS:        []string{agentstructs.SUPPORTED_OS_LINUX, agentstructs.SUPPORTED_OS_MACOS},
	},
	TaskFunctionCreateTasking: bash_executorCreateTasking,
	Version:                   1,
}

func init() {
	agentstructs.AllPayloadData.Get("freyja").AddCommand(bash_executor)
}

func bash_executorCreateTasking(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
	response := agentstructs.PTTaskCreateTaskingMessageResponse{
		Success: true,
		TaskID:  taskData.Task.ID,
	}
	if _, err := mythicrpc.SendMythicRPCArtifactCreate(mythicrpc.MythicRPCArtifactCreateMessage{
		BaseArtifactType: "ProcessCreate",
		ArtifactMessage:  "/bin/bash -c " + taskData.Args.GetCommandLine(),
		TaskID:           taskData.Task.ID,
	}); err != nil {
		logging.LogError(err, "Failed to send mythicrpc artifact create")
	}
	return response
}
